### PR TITLE
feat(enhancement): Missile Indicator Overlays

### PR DESCRIPTION
**Pull request by Koranir**
## Feature Details
This PR adds overlays to all hostile missiles on the screen. This is most useful in the beginning of the game, where the missiles deal a lot of damage but are sparse enough to dodge.

If the missile overlay is yellow, the missile is not targeting you
If the missile is targeting you it will be red-orange with spikes.
If the missile will do >10% of your HP in one hit, it will be red

This overlay is optional.

## UI Screenshots

https://user-images.githubusercontent.com/101683475/200147980-aea51ff7-3a2f-4293-bca8-7dbbef4bbedf.mp4

InAction

Yellow is missiles not targeted at you
Enemy

Orange with spikes is missiles targeted at you + Different zoom levels

https://user-images.githubusercontent.com/101683475/200147939-51d895ca-988f-410e-b192-6ff066e5c6b8.mp4

Red is a "Dangerous" missile (>5% HP damage)
Dangerous

## Performance Impact
Slightly decreased performance

### CODE REVIEW

**Comment by Koranir** [comment]
Yes, I did rip most of the code from planetLabels, why do you ask?

File: EndlessSkyLib.cbp [hunk 1]
```diff
@@ -355,12 +357,12 @@
...
 		<Unit filename="source/Weather.h" />
 		<Unit filename="source/WeightedList.h" />
+		<Unit filename="source/comparators/ByGivenOrder.h" />
+		<Unit filename="source/comparators/ByName.h" />
 		<Unit filename="source/opengl.cpp" />
 		<Unit filename="source/opengl.h" />
 		<Unit filename="source/pi.h" />
 		<Unit filename="source/shift.h" />
-		<Unit filename="source/comparators/ByGivenOrder.h" />
-		<Unit filename="source/comparators/ByName.h" />
```
Comment (RisingLeaf) [inline EndlessSkyLib.cbp:363]: Why did you change this file? seems like nothing really changed...

Comment (RisingLeaf) [reply EndlessSkyLib.cbp:363]: I mean the order
not the file

File: source/AlertLabel.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <string>
+
+class Projectile;
+class StellarObject;
+class System;
+class Ship;
+
+
+
+class AlertLabel {
```
Comment (RisingLeaf) [inline source/AlertLabel.h:32]: class descriptive comment please

File: source/AlertLabel.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+
+
+AlertLabel::AlertLabel(Point &position, const Projectile &projectile, shared_ptr<Ship> flagship, double zoom) : position(position * zoom)
+{
+
+	const bool isTargetingMe = projectile.TargetPtr() == flagship;
+	const Government *gov = projectile.GetGovernment();
+	const bool isInDanger = projectile.GetWeapon().HullDamage() > flagship->Attributes().Get("hull")/10;
```
Comment (RisingLeaf) [inline source/AlertLabel.cpp:50]: how about instead of having two states, interpolate between two states

File: EndlessSkyLib.cbp (hunk 1 above)
Comment (Koranir) [reply EndlessSkyLib.cbp:363]: It was automatic by Code::Blocks

Comment (RisingLeaf) [reply EndlessSkyLib.cbp:363]: hmmm... ok, maybe change it back? it is not on topic of this pr

File: source/AlertLabel.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	radius = zoom*projectile.Radius();
+}
+
+
+
+void AlertLabel::Draw() const
+{
+	// Draw any active labels.
+	const Font &font = FontSet::Get(14);
+	const Font &bigFont = FontSet::Get(18);
```
Comment (RisingLeaf) [inline source/AlertLabel.cpp:74]: In my opinion it does not need the text, maybe mark projectiles not targetting you blue?

File: source/AlertLabel.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+public:
+	AlertLabel(Point &position, const Projectile& projectile, std::shared_ptr<Ship> flagship, double zoom);
+
+	void Draw() const;
+
+
+private:
+	Point position;
+	double radius;
+	std::string bigText;
```
Comment (RisingLeaf) [inline source/AlertLabel.h:42]: those seem to be constant when they are set, so add them to the namespace in the cpp?
same for the colors

File: source/Engine.cpp [hunk 6]
```diff
@@ -636,6 +637,7 @@ void Engine::Step(bool isActive)
 
 	// Create the planet labels.
 	labels.clear();
+	aLabels.clear();
```
Comment (RisingLeaf) [inline source/Engine.cpp:640]: better naming please

**Comment by Amazinite** [comment]
> This PR adds overlays to all missiles on the screen.

We don't want overlays to be information overload. I think it would make more sense to only put an overlay on those missiles which are hostile to you. This is because friendly missiles aren't really important information worth highlighting.

To go a step further and reduce clutter, I'd ditch the text, and instead use a three-tiered color system. On the system radar, ships which are hostile but not actively targeting you are yellow. The same should be true with this overlay's highlighting of missiles which are hostile but not targeting you. Ships which are hostile and targeting you are then colored in orange, and so missiles which are targeting you could have their overlay be orange. I'd then reserve red for those missiles which would deal a large amount of damage to you if you were hit, regardless of whether or not they're actively targeting you. Either that, or have dangerous missiles still be colored yellow or orange depending on whether they're targeting you or not, but then gain spikes akin to the ones that appear on hostile planet labels if they're dangerous. (We could even use differing numbers of spikes to denote how powerful a missile is.)

I don't think the government color needs to be incorporated into this anywhere; the normal overlay simply distinguishes between friend and foe, and so additional overlays such as this should do the same.

File: source/Engine.cpp (hunk 6 above)
Comment (RisingLeaf) [reply source/Engine.cpp:640]: also this should probably be at the same location as the shields/hull overlays

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
Thats it for now, seems like some good code.
I like overlays but not the text, as I have stated in one comment.

File: EndlessSkyLib.cbp (hunk 1 above)
Comment (RisingLeaf) [reply EndlessSkyLib.cbp:363]: leave it like it is

**Comment by RisingLeaf** [comment]
Also some transparency on the overlays?

**Comment by RisingLeaf** [comment]
Requested review from art patrol to make sure the style is ok

**Comment by Zitchas** [comment]
a) I agree with Amazinite that I don't see a lot of value in overlays for friendly missiles. It does kind of look cool in small numbers, but quickly becomes overwhelming and useless in larger numbers.

b) I think this looks great for an Expanse style environment, as in sparse missiles that are high impact. I very much like the look; although I'm curious as to whether we could get some good looking icons instead of the words. Perhaps a lock icon or a bulls eye for the target lock; and a triangle-around-exclamation-point symbol for the danger one. I think this might look cooler, and also be more space efficient (less overlap)

While I dislike the regular ship overlays, I could see myself using these. Nice work!

**Comment by Saugia** [comment]
I'll do a more in-depth review later, but for initial thoughts, I'm going with a hard no on text. Text as an interface element in ES (as well as similar games) really only has a place on large, stationary objects in the game (planets). It doesn't belong in fast-pace combat, on ships, weapons, etc., as we rely on visual effects. Text is also not so much helpful, but more distracting. On the other hand, suggestions to have some type of visual icon may more worth your time, yes.

**Comment by Zitchas** [comment]
Regarding text - I could see text overlay being useful for identifying, say, a specific target ship, such as the main ship in a boss fight or something. Labelling the enemy flagship. But yeah, save it for special circumstances.

**Comment by Amazinite** [comment]
Just some food for thought: instead of copying from planet labels, how about a target indicator like this (imagining the red arrow as a missile, with the yellow somewhat-triangular pieces being the overlay)? Whatever the case, I think our approach should be minimalistic. As Zitchas notes, the game can get rather missile heavy at times, so the more minimalistic the missile overlay, the better.
image

**Comment by Zitchas** [comment]
I do like retaining the side logos for identifying particular dangerous munitions as well as stuff locked specifically on us; and having a logo+color combo is much more accessible (also allows it to be recognized by both text processing and visual centers of the brain, which different people react faster to.

**Comment by Zireael07** [comment]
As a glasses user, a way to spot them easier is always a good thing <3

**Comment by Koranir** [comment]
~~_**MacOS is the bane of my existence**_~~

**Comment by quyykk** [comment]
The check xcode CI creates a patch file you can apply: https://github.com/endless-sky/endless-sky/actions/runs/3392225883. If you are on Linux/MacOS you can run the `utils/check_xcode.sh` script which applies this patch automatically (if you are on Windows you can use WSL)

File: source/Engine.cpp [hunk 7]
```diff
@@ -634,6 +635,18 @@ void Engine::Step(bool isActive)
...
 		}
 
+	// Create missile overlays.
+	missileLabels.clear();
+	if(Preferences::Has("Show missile overlays"))
+	{
+		for(const Projectile &projectile : projectiles)
+		{
+			Point pos = projectile.Position() - center;
+			if((pos.Length() < Screen::Width() / zoom)  && projectile.MissileStrength() && projectile.GetGovernment()->IsEnemy())
```
Comment (RisingLeaf) [inline source/Engine.cpp:645]: why do you only test for width and not for height?

File: source/Engine.cpp [hunk 8]
```diff
@@ -634,6 +635,18 @@ void Engine::Step(bool isActive)
 			}
 		}
 
+	// Create missile overlays.
+	missileLabels.clear();
+	if(Preferences::Has("Show missile overlays"))
+	{
```
Comment (RisingLeaf) [inline source/Engine.cpp:641]: dont need the braces here

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
two small things left

**Review by Saugia — CHANGES_REQUESTED** [review]
Good that you went with yellow, orange and red for visuals. As much as faction colors would be "rule of cool" and more unique, we as a player are only human and to us, yellow orange and red are visual indicators of danger (construction equipment, etc.). 

Liking the more simplistic "outline" target shapes for these- much better than some sort of filled solid object, as most of ES's interface is broken up in pieces like that. However, judging from these images, they seem a bit too thick still- they could just about pass off as some sort of orange colored fighters. I'd suggest reducing thicknesses a bit. 

On that note, I know it's extra work, but it's especially important to get these viewed at a variety of zooms as there is no specific zoom level the playerbase overwhelmingly uses. You don't have to show all 3 colors, but if you could show at least 3 zoom levels, one all the way in, one somewhere in the middle, and farthest out, that would be very helpful.

Lastly, the forward arrow to indicate direction could just be the slightest bit longer. I am able to tell that it's there versus the other two rear points, but that's if I first focus on the missile - as it's stationary in these shots. Keep in mind these will *always* be in motion, so motion blur applies, and so does them being secondary objects in a game visual compared to other larger ones like the ships firing them.

File: source/AlertLabel.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+class Projectile;
+class StellarObject;
+class System;
+class Ship;
+
+
+// A class that holds an overlay for a missile
+class AlertLabel {
+public:
+	AlertLabel(Point &position, const Projectile& projectile, std::shared_ptr<Ship> flagship, double zoom);
```
Comment (warp-core) [inline source/AlertLabel.h:34]:
```suggestion
	AlertLabel(Point &position, const Projectile &projectile, std::shared_ptr<Ship> flagship, double zoom);
```
Do we need to modify `position`? Or, can it be const?
Should the `flagship` pointer be taking a reference? Should that be const?

File: source/AlertLabel.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+private:
+	double rotation;
+	Point position;
+	double zoom;
+	bool isDangerous;
+	bool isTargetingMe;
+	double radius;
+	float damagePercent;
+	Color color;
```
Comment (warp-core) [inline source/AlertLabel.h:47]: Initialise primitives.
Should `color` be a pointer?
```suggestion
	double rotation = 0.;
	Point position;
	double zoom = 1.;
	bool isDangerous = false;
	bool isTargetingMe = false;
	double radius;
	float damagePercent = 0.;
	Color color;
```
I don't know what the default for `radius` should be.

File: source/AlertLabel.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+
+
+AlertLabel::AlertLabel(Point &position, const Projectile &projectile, shared_ptr<Ship> flagship, double zoom) : position(position * zoom), zoom(zoom)
+{
+
+	isTargetingMe = projectile.TargetPtr() == flagship;
+	isDangerous = projectile.GetWeapon().HullDamage() > flagship->Attributes().Get("hull")/10;
+	damagePercent = projectile.GetWeapon().HullDamage() / flagship->Attributes().Get("Hull");
```
[the remark this answers is not in the document]
Comment (Amazinite) [reply source/AlertLabel.cpp:46]: I do think it doesn't make much sense to mark a missile as dangerous if it does a lot of hull damage but your shields are up.

**Comment by Amazinite** [comment]
The solid red circle around "dangerous" missiles still seems too much clutter to me. When I said we should use color to denote danger level, part of the goal of that was that missiles which the player should pay more attention to should be primarily marked by color, not be significantly increased screen real estate.

File: source/AlertLabel.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+	if(!isTargetingMe)
+		color = Color(.8f, .8f, .4f, .5f);
+	else
+		color = Color(5.f, .48f, .16f, .5f);
+
+	if(isDangerous)
+		color = Color(1.f, .6f, .4f, .5f);
```
[the remark this answers is not in the document]
Comment (Amazinite) [reply source/AlertLabel.cpp:54]: For future reference, all new Colors should be defined in interfaces.txt so that they can be tweaked to the user's preference.

File: source/AlertLabel.cpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+AlertLabel::AlertLabel(const Point position, const Projectile &projectile, const shared_ptr<Ship> flagship, const double zoom)
+		: position(position * zoom), zoom(zoom)
+{
+	isTargetingMe = projectile.TargetPtr() == flagship;
+	double maxHP = flagship->Attributes().Get("hull") + flagship->Attributes().Get("shield");
+	double missileDamage = projectile.GetWeapon().HullDamage() + projectile.GetWeapon().ShieldDamage();
+	damagePercent = missileDamage / maxHP;
+	isDangerous = damagePercent > DANGEROUS_ABOVE;
+
+	if(!isTargetingMe)
```
Comment (quyykk) [inline source/AlertLabel.cpp:51]: Switch the condition around, i.e. `if(isTargetingMe)` because you have an else anyways. You can also do an if else chain if you move the isDangerouns `if` first.

File: source/AlertLabel.cpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+using namespace std;
+
+namespace {
+	const double DANGEROUS_ABOVE = .05;
+}
+
+
+
+AlertLabel::AlertLabel(const Point position, const Projectile &projectile, const shared_ptr<Ship> flagship, const double zoom)
```
Comment (quyykk) [inline source/AlertLabel.cpp:42]:
```suggestion
AlertLabel::AlertLabel(const Point &position, const Projectile &projectile, const shared_ptr<Ship> &flagship, double zoom)
```

File: source/AlertLabel.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+*/
+
+#ifndef ALERT_LABEL_H_
+#define ALERT_LABEL_H_
+
+#include "Color.h"
+#include "Point.h"
+
+#include <memory>
+#include <string>
```
Comment (quyykk) [inline source/AlertLabel.h:23]:
```suggestion
```

unused include

File: source/AlertLabel.h [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+class Projectile;
+class StellarObject;
```
Comment (quyykk) [inline source/AlertLabel.h:26]:
```suggestion
```

File: source/AlertLabel.h [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+
+private:
+	double rotation = 0;
+	Point position;
+	double zoom = 1.;
+	bool isDangerous = false;
+	bool isTargetingMe = true;
+	double radius = 15.;
+	float damagePercent = 0.;
```
Comment (Amazinite) [inline source/AlertLabel.h:46]: This value is only used to determine if isDangerous is true. Why hold onto it?

File: source/AlertLabel.h [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+	void Draw() const;
+
+
+private:
+	double rotation = 0;
+	Point position;
+	double zoom = 1.;
+	bool isDangerous = false;
+	bool isTargetingMe = true;
```
Comment (Amazinite) [inline source/AlertLabel.h:44]: I'd rename this to isTargetingPlayer or isTargetingFlagship.

File: source/AlertLabel.cpp (hunk 3 above)
Comment (Koranir) [reply source/AlertLabel.cpp]: Outdated, but hey are now separate flags that can be active independently.

**Comment by Saugia** [comment]
Slightly confused, is there a reason some of your images in the PR have the pointers around the graphics and others don't?

**Comment by Koranir** [comment]
> Slightly confused, is there a reason some of your images in the PR have the pointers around the graphics and others don't?

When the missiles lock onto the flagship they gain pointy bits and change colour from yellow to orange, but the colour can be overriden to red if the missile is dangerous.

**Review by Saugia — APPROVED** [review]
Looking good on the art/display side of things. Will defer this to a code dev again to finalize that part, then see where we sit.

**Review by RisingLeaf — APPROVED** [review]

File: source/Engine.cpp [hunk 19]
```diff
@@ -634,6 +635,16 @@ void Engine::Step(bool isActive)
 			}
 		}
 
+	// Create missile overlays.
+	missileLabels.clear();
+	if(Preferences::Has("Show missile overlays"))
+		for(const Projectile &projectile : projectiles)
+		{
+			Point pos = projectile.Position() - center;
+			if((pos.Length() < max(Screen::Width(), Screen::Height()) * .5 / zoom)  && projectile.MissileStrength() && projectile.GetGovernment()->IsEnemy())
```
Comment (Amazinite) [inline source/Engine.cpp:644]: Take advantage of short-circuiting and put the most expensive calculation last.
```suggestion
			if(projectile.MissileStrength() && projectile.GetGovernment()->IsEnemy() && (pos.Length() < max(Screen::Width(), Screen::Height()) * .5 / zoom))
```

File: source/AlertLabel.h [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+// A class that holds an overlay for a missile
+class AlertLabel {
+public:
+	AlertLabel(const Point &position, const Projectile &projectile, const std::shared_ptr<Ship> &flagship, double zoom);
+
+	void Draw() const;
+
+
+private:
+	double rotation = 0;
```
Comment (Amazinite) [inline source/AlertLabel.h:37]:
```suggestion
	double rotation = 0.;
```

File: source/AlertLabel.h [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+#include "Color.h"
+#include "Point.h"
+
+#include <memory>
+
+class Projectile;
+class Ship;
+
+
+// A class that holds an overlay for a missile
```
Comment (Amazinite) [inline source/AlertLabel.h:28]:
```suggestion
// A class that holds an overlay for a missile.
```

File: source/AlertLabel.cpp [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+
+void AlertLabel::Draw() const
+{
+	double angle[3] = {330., 210., 90.};
+	for(int i = 0; i < 3; i++)
+	{
+		RingShader::Draw(position * zoom, radius, 1.2f, .16f, *color, 0.f, angle[i] + rotation);
+		if(isTargetingFlagship)
+			PointerShader::Draw(position * zoom, Angle(angle[i] + 30. + rotation).Unit(), 7.5f, (i ? 10.f : 22.f)*zoom, radius + (i ? 10 : 20)*zoom, *color);
```
Comment (Amazinite) [inline source/AlertLabel.cpp:70]:
```suggestion
			PointerShader::Draw(position * zoom, Angle(angle[i] + 30. + rotation).Unit(),
				7.5f, (i ? 10.f : 22.f) * zoom, radius + (i ? 10.f : 20.f) * zoom, *color);
```

**Review by Amazinite — COMMENTED** [review]
Just a quick style check.

**Review by Amazinite — APPROVED** [review]

**Review by quyykk — APPROVED** [review]
Nice thanks!
